Odata-docs/webapi-8/tutorials/custom-skiptokenhandler.md

```diff
@@ -0,0 +1,549 @@
+
+```http
+GET /odata/Customers
+```
+
+**Response**:
+
+```json
+**Response**:
+
+```json
+{
+    "@odata.context": "http://localhost:5000/odata/$metadata#Customers",
+    "value": [
+        {
+            "Id": 1,
+            "Name": "Customer 1"
+        },
+        {
+            "Id": 2,
+            "Name": "Customer 2"
+        }
+    ],
+    "@odata.nextLink": "http://localhost:5000/odata/Customers?$skiptoken=SWQtMg%3d%3d"
+}
+```
+
+The `$skiptoken` is now `SWQtMg%3d%3d` instead of `Id-2`.
+
+> [!NOTE]
+> The actual base64 encoding of `Id-2` is `SWQtMg==`, but the final form eventually gets URL encoded to `SWQtMg%3d%3d`.
+
+Let's follow the next link to fetch the next page:
+
+```http
+GET http://localhost:5000/odata/Customers?$skiptoken=SWQtMg%3d%3d
+```
+
+This returns an error response like the following:
+
+```json
+{
+    "error": {
+        "code": "",
+        "message": "The query specified in the URI is not valid. Unable to parse the skiptoken value 'SWQtMg=='. Skiptoken value should always be server generated.",
+    }
+}
+```
+
+The reason we get the error is because our custom handler was not able to process the skip token. The skip token is processed in the `ApplyTo` method. At the moment, we are still using the `ApplyTo` method in the `DefaultSkipTokenHandler` class because we have not defined our own. The base `ApplyTo` expects the skip token to be in a specific format, it cannot parse the base64 encoded value.
+
+### Processing the skip token with `ApplyTo`
+
+To fix the error in the previous section, we have to decode the skip token first then pass the decoded version to the default skip token handler.
+
+We can achieve that by adding the following methods to our `CustomSkipTokenHandler` class:
+
+```csharp
+public override IQueryable<T> ApplyTo<T>(IQueryable<T> query, SkipTokenQueryOption skipTokenQueryOption, ODataQuerySettings querySettings, ODataQueryOptions queryOptions)
+{
+    var decodedSkipTokenQueryOption = DecodeSkipToken(skipTokenQueryOption);
+    return base.ApplyTo(query, decodedSkipTokenQueryOption, querySettings, queryOptions);
+}
+
+public override IQueryable ApplyTo(IQueryable query, SkipTokenQueryOption skipTokenQueryOption, ODataQuerySettings querySettings, ODataQueryOptions queryOptions)
+{
+    var decodedSkipTokenQueryOption = DecodeSkipToken(skipTokenQueryOption);
+    return base.ApplyTo(query, decodedSkipTokenQueryOption, querySettings, queryOptions);
+}
+
+private static SkipTokenQueryOption DecodeSkipToken(SkipTokenQueryOption skipTokenQueryOption)
+{
+    string encodedSkipToken = skipTokenQueryOption.RawValue;
+    string decodedSkipToken = Encoding.UTF8.GetString(Convert.FromBase64String(encodedSkipToken));
+    return new SkipTokenQueryOption(decodedSkipToken, skipTokenQueryOption.Context);
+}
+```
+
+We override the `ApplyTo` methods with our custom logic. We override both the generic `ApplyTo` and generic `ApplyTo<T>` overloads. The methods accept the following parameters:
+
```

Comment:
```suggestion
```

Reply:
The [Learn Authoring Pack](https://marketplace.visualstudio.com/items?itemName=docsmsft.docs-authoring-pack) extension issues a warning if you don't place a blank line before a list.